### ISSUE TITLE
Inertia::merge() Improperly Merges Pagination Objects in Inertia v2

### ISSUE THREAD

**Issue by HichemTab-tech** [inertiajs/inertia#2068]
### Version:

- `@inertiajs/react` version: 2.0.0-beta.1

### Describe the problem:

In Inertia v2, the new `merge` feature allows developers to merge new data into existing props instead of replacing them. However, there is an issue with how arrays that are nested within objects are being handled. Specifically, when an incoming prop is an object containing an array (e.g., pagination data), the array is not being merged properly.

#### Current Behavior:
- In the source code:
  ```javascript
  if (Array.isArray(incomingProp)) {
      pageResponse.props[prop] = [...((currentPage.get().props[prop] || []) as any[]), ...incomingProp]
  } else if (typeof incomingProp === 'object') {
      pageResponse.props[prop] = {
          ...((currentPage.get().props[prop] || []) as Record<string, any>),
          ...incomingProp,
      }
  }
  ```
_You can check it here : [source](https://github.com/inertiajs/inertia/blob/9106d74495130a3d82d00daec43c4042b81bdb63/packages/core/src/response.ts#L246C9-L253C10)_

- If the `incomingProp` is an array, it is merged properly by appending the new elements to the existing ones.
- If the `incomingProp` is an object, the properties are shallowly merged. This means that if the object contains an array (e.g., a pagination object with `data`, `meta`, and `links`), the array (`data`) is not merged correctly. Instead, it is replaced.

For example, consider a pagination object with the following structure:
```javascript
{
    data: [/* items */],
    meta: { /* pagination metadata */ },
    links: { /* pagination links */ }
}
```

If you attempt to merge this object using `Inertia::merge()`, the `data` array will not be merged (appended to). Instead, it will be replaced by the new array, which leads to the loss of existing pagination items.

#### Expected Behavior:
- When using `Inertia::merge()`, arrays inside objects (e.g., `data` within a pagination object) should also be merged, similar to how standalone arrays are merged.
- Specifically, `data` within a pagination object should be appended to rather than replaced, while other properties (`meta`, `links`) should be replaced or handled appropriately.

#### Steps to Reproduce:
1. Use `Inertia::merge()` in the backend to pass a pagination object as a prop.
2. Observe how the `data`, `meta`, and `links` properties are merged in the frontend.
3. Notice that while standalone arrays are merged correctly, the `data` array within the pagination object is replaced rather than appended.

#### Possible Solution:
- Implement a deep merging strategy for nested arrays within objects. For example, if an object contains an array property, that array should be merged rather than replaced.
- Alternatively, allow developers to define custom merge behaviors for nested properties to prevent unintended replacements.

**Comment by YounesAmalou** [comment]
Interesting observation, merge should actually merge the actual `data` object rather replacing it,

For example this might cause issues with paging if we try to send an object that contains the item list as an array + some meta data such as the max page, current page, next page...

Their might be other scenarios where the merge is actually needed.

This might not be convenient for people who expect to merge the actual data while not being familiar with this behavior,

I would suggest making this behaviors clear in the documentation while adding support for another method that handles the actual merge to keep both functions in case we actually need to replace the response.

**Comment by rugaard** [comment]
Seems to be the same issue or very similar in the Vue version as well.

**Comment by HichemTab-tech** [comment]
> Seems to be the same issue or very similar in the Vue version as well.

@rugaard yes seems to be related to the inertia core, well I suggested a solution for this in these PRs ([inertia-laravel](https://github.com/inertiajs/inertia-laravel/pull/679) & [](https://github.com/inertiajs/inertia/pull/2069)) where I introduced `deepMerge`, I hope they take a look at it soon.

**Comment by JasperTey** [comment]
I agree that this is a valid issue, and wanted to add my support for it. The examples shown in the docs: https://inertiajs.com/merging-props#server-side - would currently NOT work since they all return `User::paginate()` responses.

**Comment by MohammadZarifiyan** [comment]
Any updates?

**Comment by HichemTab-tech** [comment]
> Any updates?

Not yet unfortunately :"/ .

**Comment by dillingham** [comment]
maybe @joetannenbaum can save us

**Comment by HichemTab-tech** [comment]
Hey @joetannenbaum , can you take  a look at this please?

**Comment by tauseefsshah** [comment]
Hey @joetannenbaum can you please look into this. Infinite scroll and similar features that require merging pagination props aren't working 😅
This will be a big help, Thanks.

**Comment by joetannenbaum** [comment]
Hi everyone. I will check into this soon. I'm parsing through a lot of issues and PRs right now, but this is indeed on my radar. 

Just for some clarity: this was intentionally a shallow merge for now, with plans to enhance later. So it's currently working as expected, but I hear you that it needs to be altered to work with a deeper merge.

**Comment by dillingham** [comment]
> Hi everyone. I will check into this soon. I'm parsing through a lot of issues and PRs right now, but this is indeed on my radar. 
> 
> Just for some clarity: this was intentionally a shallow merge for now, with plans to enhance later. So it's currently working as expected, but I hear you that it needs to be altered to work with a deeper merge. 

The v2 docs show merging laravel pagination which does not work..

**Comment by HichemTab-tech** [comment]
> Hi everyone. I will check into this soon. I'm parsing through a lot of issues and PRs right now, but this is indeed on my radar.
> 
> Just for some clarity: this was intentionally a shallow merge for now, with plans to enhance later. So it's currently working as expected, but I hear you that it needs to be altered to work with a deeper merge.

Thanks for the update, I understand the current behavior was intentional. I’ve already proposed a PR for deeper merge functionality and mentioned it in the issue. Would love for you to check it.

**Comment by HichemTab-tech** [comment]
Hey @joetannenbaum ! any updates?

**Comment by MohammadZarifiyan** [comment]
Hey @joetannenbaum ! 
We are looking forward to this update.
Please let us know if there is a new update.

**Comment by MohammadZarifiyan** [comment]
This feature is not working correctly on `@inertiajs/vue`

**Comment by Trevor57** [comment]
Hello everyone, what's new on this ?

**Comment by Trevor57** [comment]
@joetannenbaum any updates?

**Comment by kaiserkiwi** [comment]
Before I get crazy: Is it correct, that something like that doesn't work currently on the **Vue**-Version?

```php
'posts' => Inertia::merge(Post::orderBy('published_at', 'desc')->paginate()),
```

I get the `mergeProps: ["posts"]` in my response object, but the Frontend seems to ignore it completely.

**Comment by Trevor57** [comment]
> Folks give Joe some time, I'm sure he's caught up with a lot. You can use the following solution for pagination in the meantime:
> 
> $courses = Course::paginate(5);
> 
> return Inertia:: render( 'Courses/Index', [
>     // For data items
>     'courses' => Inertia: :merge($courses-›items()),
>     // For pagination items
>     'pagination' => Arr:: except($courses-›toArray(), ['data']),
> );
> This solution was shared by Jeffrey Way on [X](https://x.com/jeffrey_way/status/1885042229326340563)

Thanks a lot, it kind of unblock us a little, but still the merging doesn't work with pagination when using resources api :"(

**Comment by tauseefsshah** [comment]
> the merging doesn't work with pagination when using resources api

@Trevor57 Can you please explain more or share some code snippet, I'll try to help :)
You can also try posting in the laravel discord `#help` channel

**Comment by deondazy** [comment]
> @Trevor57 the merging doesn't work with pagination when using resources api :"(

I got infinite scroll to work when using resources api. [See this post](https://dev.to/deondazy/how-to-combine-filters-sorting-and-infinite-scrolling-in-laravel-inertiajs-v2-and-vue-3-24a7)

**Comment by kaiserkiwi** [comment]
> but still the merging doesn't work with pagination when using resources api

Can kind of confirm. It works here without a resource but the moment I add one it only works if `preserveUrl` is set to `false`.

Example 
```vue
<WhenVisible
	:always="pagination.current_page < pagination.last_page"
	:buffer="100"
	:params="{
		data: {
			page: pagination.current_page + 1,
		},
		only: ['pagination', 'items'],
		preserveUrl: true,
	}" />
```

```php
return Inertia::render('Items/Index', [
	'items' => Inertia::merge(fn () => ItemResource::collection($items->items())),
	'pagination' => Arr::except($$items->toArray(), ['data']),
]);
```

This doesn't work. 
But if I change `preserveUrl: true,` to `preserveUrl: false,` it will work. But I don't want the pagination in the URL in my case.

If I change the PHP to the following it works too:
```php
return Inertia::render('Items/Index', [
	'items' => Inertia::merge(fn () => $items->items()),
	'pagination' => Arr::except($$items->toArray(), ['data']),
]);
```

But also: I don't want to have the whole object here.

**Comment by rori-tallbert41** [comment]
Hello everyone, I hope you're doing great, just wanted to know is there any updates on this isssue ?

**Comment by rori-tallbert41** [comment]
Hello @joetannenbaum can you check on this issue, please ?

**Comment by jonnywilliamson** [comment]
Yeah this is starting to affect me too with trying to launch a new site that needs an infinite scroll. I know it's been an insane busy week for Laravel releases and do not want to sound demanding in the slightest, but it would be fantastic to see this issue patched up!

**Comment by safarovitch** [comment]
https://laracasts.com/series/inertia-2-unleashed/episodes/5

Temporary fix. Watch from 6:40.

**Comment by CodeWrap** [comment]
For those who want to use API Resources:

```
$items = Item::query()->paginate();

$resource = ItemResource::collection($items);

return Inertia::render('Items/Index', [
    'items' => Inertia::merge(json_decode($resource->toJson(), true)),
    'itemsMeta' => Arr::except($items->toArray(), 'data'),
]);
```

**Comment by actengage** [comment]
Hopefully this thread gets some attention soon. I wanted to add my take. It would be useful for the Inertia API to have the ability to implement a custom merge strategy. I highly doubt a one-size-fits all approach would work in all cases. In my case, I am doing some groupby's on a flat collection that organizes threaded responses as emails come. Consider the Outlook web UI for example. Emails are grouped by a time frame, then into a conversation thread, and then an array of messages in a thread.

Consider the following. I have a deferred prop that loads a structure with the following type:

```
{
    groups: Record<string, MailboxMessage[][]>;
    page: int;
    perPage: int;
    timezone: string;
}
```

This results in some data looking so:

```
groups: {
    "Today": [
        [
            {"id": 2, "subject": "Re: Some Email Subject"},
            {"id": 1, "subject": "Some Email Subject"}
        ],
        [
            {"id": 4, "subject": "Re: Hello!"},
            {"id": 3, "subject": "Hello!"}
        ]
    ],
    "Yesterday": [
        [
            {"id": 6, "subject": "Re: Some Email Subject"},
            {"id": 5, "subject": "Some Email Subject"}
        ]
    ]
}
```

I haven't looked into the internals of the package to see how merges are currently handled or if this API possible, but perhaps something like this. The idea is that we get a callback to handle a merge with our own strategy.

```
'groups' => Inertia::defer(function () use ($request, $mailbox, $folder): ConversationGroups {
    return app(MailboxMessageService::class)->groups(
        mailbox: $mailbox,
        folder: $folder,
        timezone: Auth::user()->timezone(),
        order: $request->order(),
        sort: $request->sort(),
        page: $request->get('page', 1),
        perPage: $request->get('per_page', 20),
    );
}, 'groups')->mergeWith(function($newData, $oldData) {
   return myCustomMergeStrategy($newData, $oldData);
})
```

I wouldn't expect this to be able to be merge out of the box (although if its possible, thats great), but I would expect a way to be able to implement my own merge strategy. The best I can figure at this time, without jumping through a bunch of hoops with a bunch of additional inertia props, the only practical solution for this would be just handle all the grouping logic in the JS components instead of server side. However, Laravel makes all this much easier with collections and I'd prefer it to be handled server side.

I'd be interesting in anyone else's feedback on a possible approach. Am I overthinking this or is this just a bad approach?

**Comment by actengage** [comment]
I just wanted to follow up on this thread. I dug into the laravel and JS side of inertia and it appears the merge is done client side (I could be wrong). But i couldn't find anywhere that indicated this was a server-side process. Maybe my proposed solution just isnt possible and the wrong approach. However after digging into this more, the solution was to not use merge in inertia and just do it client side. 

I do wish the documentation was more verbose , with better example, and I think this core bug should be fixed. It should be clear how things are merged and work, with better ways to control it, however that may be.

However, it wasn't hard to achieve this manually once I understood more how it worked. In my case, I'm trying to use infinite scrolling without changing my data structure on the server.

This example uses Vue.js.

```ts
const data = ref<ConversationGroups>();
const hasMorePages = ref(true);
const isLoading = ref(false);

function mergeGroupsProp(groups: ConversationGroups) {
    // my logic for merging the props and determine if there are more pages the load without using a LengthAwarePaginator.
}

watchEffect(() => {
    if(props.groups) {
        mergeGroupsProp(props.groups);
    }
});
```

Then the when visible component. Hopefully this gives some insight for possible workarounds without polluting the controller with a bunch of props, while using lazy/deferred props.

```
<WhenVisible
    v-if="hasMorePages"
    :always="true"
    :params="{
        preserveUrl: true,
        only: ['groups'],
        data: {
            page: data?.page + 1
        },
        onStart: () => isLoading = true,
        onFinish: () => isLoading = false,
    }" />
```

**Comment by HichemTab-tech** [comment]
Hello @actengage, so the process here is that the merging is indeed happening on the frontend side — more precisely [here](https://github.com/inertiajs/inertia/blob/9106d74495130a3d82d00daec43c4042b81bdb63/packages/core/src/response.ts#L246C9-L253C10). On the backend, we’re not doing much more than just indicating which props need to be merged and which don’t (you can check that [here](https://github.com/inertiajs/inertia-laravel/blob/32e80145ae36e89742b90a654ec482911a0bd1b7/src/ResponseFactory.php#L120) and [here](https://github.com/inertiajs/inertia-laravel/blob/32e80145ae36e89742b90a654ec482911a0bd1b7/src/Response.php#L319)).

So yeah, we can’t directly define a merging strategy on the backend, since the actual merging is handled by the JS frontend. **But** you're right — we *could* make something happen by establishing some sort of common structure between backend and frontend. Not necessarily a strict schema, but maybe a flexible "language" or convention — just enough to describe what and how to merge, without being too open-ended or rigid.

Here’s what I had in mind:

```php
$object = [
   "data" => [...],
   "other-array" => [...],
];
```

Let’s say we only want to merge the `data` key — not all arrays (which `deepMerge`, introduced in the related PR, should do). We could do something like:

```php
Inertia::merge($object, "data");
```

Or maybe, if we can define some merge behavior:

```php
Inertia::merge($object, ["data" => MERGE_TYPE1, "other-array" => MERGE_TYPE2]);
```

See? We *can* do something custom — but **only** if we have some common rules or structure shared between backend and frontend, so that the frontend knows how to handle the result based on those rules.

But for now i thought we can start by adding a general `deepMerge`, and for these thoughts may comme after :).

**Comment by actengage** [comment]
Thanks for the reply. I really appreciate it. I think the concept is pretty good. I agree, just a deepMerge strategy would be the most obvious place to start. Thinking about the more advanced concept you proposed, I think some sensible structures could make sense. 

But this makes me think a bit about how a more flexible pattern could be implemented based on this concept. I wonder how a concept would work if something like the following was introduced. This would allow merge logic to be defined once in a global place like app.ts.

```ts
import { defineMerge } from '@inertiajs/core';

defineMerge('some_merge_stragegy', (some, props, here) => {
  return myCustomMergeStrategy(some, props, here);
});
```

And with PHP you could use the key defined above to trigger whatever merge strategy one desires. If the key isn't defined, throw a console warning. In fact MERGE_TYPE1 or MERGE_TYPE2 could use the same underlying API.

```php
Inertia::merge($object, ['data' => 'some_merge_strategy"]);
```

**Comment by HichemTab-tech** [comment]
Well yes that's a really good idea too, i mean we can have our own custom merge strategy without really defining anything in the backend just the name of the strategy.

It would be great if we have that.
